homeassistant/components/nest/config_flow.py

```diff
@@ -47,22 +62,57 @@ class CodeInvalid(NestAuthError):
 
 
 @config_entries.HANDLERS.register(DOMAIN)
-class NestFlowHandler(config_entries.ConfigFlow):
-    """Handle a Nest config flow."""
+class NestFlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
+    """Config flow to handle authentication for both APIs."""
 
+    DOMAIN = DOMAIN
     VERSION = 1
     CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_PUSH
 
-    def __init__(self):
-        """Initialize the Nest config flow."""
-        self.flow_impl = None
+    @classmethod
+    def register_sdm_api(cls, hass):
+        """Configure the flow handler to use the SDM API."""
+        if DOMAIN not in hass.data:
+            hass.data[DOMAIN] = {}
+        hass.data[DOMAIN][DATA_SDM] = {}
+
+    def is_sdm_api(self):
+        """Return true if this flow is setup to use SDM API."""
+        return DOMAIN in self.hass.data and DATA_SDM in self.hass.data[DOMAIN]
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> Dict[str, str]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {
+            "scope": " ".join(SDM_SCOPES),
+            # Add params to ensure we get back a refresh token
+            "access_type": "offline",
+            "prompt": "consent",
+        }
+
+    async def async_oauth_create_entry(self, data: dict) -> dict:
+        """Create an entry for the SDM flow."""
+        data[DATA_SDM] = {}
+        return await super().async_oauth_create_entry(data)
 
     async def async_step_user(self, user_input=None):
         """Handle a flow initialized by the user."""
+        if self.is_sdm_api():
+            return await super().async_step_user(user_input)
         return await self.async_step_init(user_input)
 
     async def async_step_init(self, user_input=None):
         """Handle a flow start."""
+        if self.is_sdm_api():
+            return None
```

Comment:
This would raise a TypeError in the flow manager. Is that intended?

https://github.com/home-assistant/core/blob/c677489535be5d44efaf6baa01a96f2dddd4a1af/homeassistant/data_entry_flow.py#L214